lib/service/awsoidc.go

```diff
@@ -0,0 +1,263 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package service
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"golang.org/x/time/rate"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/utils/retryutils"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/automaticupgrades"
+	"github.com/gravitational/teleport/lib/integrations/awsoidc"
+	"github.com/gravitational/teleport/lib/utils/interval"
+)
+
+const (
+	// updateDeployAgentsInterval specifies how frequently to check for available updates.
+	updateDeployAgentsInterval = time.Minute * 30
+
+	// updateDeployAgentsRateLimit specifies the time between updates across AWS regions.
+	updateDeployAgentsRateLimit = time.Second * 30
+)
+
+func (process *TeleportProcess) periodUpdateDeployServiceAgents() error {
+	if !process.Config.Proxy.Enabled {
+		return nil
+	}
+
+	// start process only after teleport process has started
+	if _, err := process.WaitForEvent(process.GracefulExitContext(), TeleportReadyEvent); err != nil {
+		return nil
+	}
+	process.log.Infof("The new service has started successfully. Checking for deploy service updates every %v.", updateDeployAgentsInterval)
+
+	resp, err := process.getInstanceClient().Ping(process.GracefulExitContext())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if !resp.ServerFeatures.AutomaticUpgrades {
+		return nil
+	}
+
+	periodic := interval.New(interval.Config{
+		Duration: updateDeployAgentsInterval,
+		Jitter:   retryutils.NewSeventhJitter(),
+	})
+	defer periodic.Stop()
+
+	for {
+		if err := process.updateDeployServiceAgents(process.GracefulExitContext(), process.getInstanceClient()); err != nil {
+			process.log.Warningf("Update failed: %v. Retrying in ~%v", err, updateDeployAgentsInterval)
+		}
+
+		select {
+		case <-periodic.Next():
+		case <-process.GracefulExitContext().Done():
+			return nil
+		}
+	}
+}
+
+func (process *TeleportProcess) updateDeployServiceAgents(ctx context.Context, authClient *auth.Client) error {
```

Comment:
I don't know if this should be a part of `TeleportProcess` to be honest.

Can we make a proper separate "service struct" e.g. `FargateUpdater` or something in lib/automaticupgrades, that would get initialized with a config struct, implement all necessary upgrade business logic (basically, this method) and provide a proper interface to Teleport process to call?